[1][Span logging] Introduce DB migrations for SqlAlchemyStore span logging

dbczumar opened the pull request:
<details><summary>&#x1F6E0 DevTools &#x1F6E0</summary>
<p>

[![Open in GitHub Codespaces](https://github.com/codespaces/badge.svg)](https://codespaces.new/dbczumar/mlflow/pull/17073?quickstart=1)

#### Install mlflow from this PR

```
# mlflow
pip install git+https://github.com/mlflow/mlflow.git@refs/pull/17073/merge
# mlflow-skinny
pip install git+https://github.com/mlflow/mlflow.git@refs/pull/17073/merge#subdirectory=libs/skinny
```

For Databricks, use the following command:

```
%sh curl -LsSf https://raw.githubusercontent.com/mlflow/mlflow/HEAD/dev/install-skinny.sh | sh -s pull/17073/merge
```

</p>
</details>

- Add migration a1b2c3d4e5f6 to create spans table
- Add SqlSpan model to dbmodels/models.py
- Include indexes for efficient querying by experiment_id, status, type, and duration
- Remove trace_state column from schema (never included in final migration)

### Related Issues/PRs

<!-- Uncomment 'Resolve' if this PR can close the linked items. -->
<!-- Resolve --> #xxx

### What changes are proposed in this pull request?

Introduce DB migrations for SqlAlchemyStore span logging

### How is this PR tested?

- [X] Existing unit/integration tests
- [ ] New unit/integration tests
- [ ] Manual tests

<!-- Attach code, screenshot, video used for manual testing here. -->

### Does this PR require documentation update?

- [X] No. You can skip the rest of this section.
- [ ] Yes. I've updated:
  - [ ] Examples
  - [ ] API references
  - [ ] Instructions

### Release Notes

#### Is this a user-facing change?

- [X] No. You can skip the rest of this section.
- [ ] Yes. Give a description of this change to be included in the release notes for MLflow users.

<!-- Details in 1-2 sentences. You can just refer to another PR with a description if this PR is part of a larger change. -->

#### What component(s), interfaces, languages, and integrations does this PR affect?

Components

- [X] `area/tracking`: Tracking Service, tracking client APIs, autologging
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/evaluation`: MLflow model evaluation features, evaluation metrics, and evaluation workflows
- [ ] `area/gateway`: MLflow AI Gateway client APIs, server, and third-party integrations
- [ ] `area/prompts`: MLflow prompt engineering features, prompt templates, and prompt management
- [ ] `area/tracing`: MLflow Tracing features, tracing APIs, and LLM tracing functionality
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/docs`: MLflow documentation pages

<!--
Insert an empty named anchor here to allow jumping to this section with a fragment URL
(e.g. https://github.com/mlflow/mlflow/pull/123#user-content-release-note-category).
Note that GitHub prefixes anchor names in markdown with "user-content-".
-->

<a name="release-note-category"></a>

#### How should the PR be classified in the release notes? Choose one:

- [X] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [ ] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

#### Should this PR be included in the next patch release?

`Yes` should be selected for bug fixes, documentation updates, and other small changes. `No` should be selected for new features and larger changes. If you're unsure about the release classification of this PR, leave this unchecked to let the maintainers decide.

<details>
<summary>What is a minor/patch release?</summary>

- Minor release: a release that increments the second part of the version number (e.g., 1.2.0 -> 1.3.0).
  Bug fixes, doc updates and new features usually go into minor releases.
- Patch release: a release that increments the third part of the version number (e.g., 1.2.0 -> 1.2.1).
  Bug fixes and doc updates usually go into patch releases.

</details>

<!-- patch -->

- [ ] Yes (this PR will be cherry-picked and included in the next patch release)
- [X] No (this PR will be included in the next minor release)

dbczumar committed:
Add database migration and model for spans table

- Add migration a1b2c3d4e5f6 to create spans table
- Add SqlSpan model to dbmodels/models.py
- Include indexes for efficient querying by experiment_id, status, type, and duration
- Remove trace_state column from schema (never included in final migration)

Signed-off-by: dbczumar <corey.zumar@databricks.com>

dbczumar committed:
Add comment explaining MySQL LONGTEXT usage for span content

Explain that LONGTEXT is needed for MySQL to support large span content
up to 4GB, as standard TEXT is limited to 64KB which is insufficient
for spans with extensive attributes, events, or nested data

Signed-off-by: dbczumar <corey.zumar@databricks.com>

dbczumar committed:
Fix MSSQL index error by changing type column from TEXT to VARCHAR

MSSQL doesn't allow TEXT columns in indexes, which causes an error when
creating composite indexes that include the type column. Changed type
column from sa.Text() to sa.String(200) to support MSSQL while still
providing ample space for span type values.

Signed-off-by: dbczumar <corey.zumar@databricks.com>

dbczumar committed:
Increase span type column length to 1000 characters

Increase the type column limit from 200 to 1000 characters to provide
more flexibility for span type values while still supporting MSSQL indexes

Signed-off-by: dbczumar <corey.zumar@databricks.com>

dbczumar committed:
Update comments to note 1000 chars should be plenty for span types

Add clarification in both migration and model comments that 1000
characters should be sufficient for most span type identifiers

Signed-off-by: dbczumar <corey.zumar@databricks.com>

dbczumar committed:
Update schema snapshot to include spans table

Add spans table definition to latest_schema.sql with correct column types:
- name as TEXT
- type as VARCHAR(1000) for MSSQL compatibility
- duration_ns as computed column
- Removed trace_state column that was removed from design

Signed-off-by: dbczumar <corey.zumar@databricks.com>

dbczumar commented on tests/resources/db/latest_schema.sql:
```diff
@@ -269,6 +269,24 @@ CREATE TABLE params (
+	status VARCHAR(50) NOT NULL,
+	start_time_unix_nano BIGINT NOT NULL,
+	end_time_unix_nano BIGINT,
+	duration_ns BIGINT GENERATED ALWAYS AS (end_time_unix_nano - start_time_unix_nano) STORED,
```
For efficient filtering on latency

dbczumar replied:
Yes :)

B-Step62 reviewed (approved):
The schema looks good overall, but it seems the database test is failing.

dbczumar committed:
fix

Signed-off-by: dbczumar <corey.zumar@databricks.com>

dbczumar committed:
Reduce span type column to 500 chars for MySQL index compatibility

MySQL has a max index key length of 3072 bytes. With utf8mb4 encoding
(4 bytes per char), the composite index on (experiment_id, status, type)
was exceeding this limit with type VARCHAR(1000). Reduced to 500 chars
which keeps us well within the limit while still providing ample space
for span type identifiers.

Signed-off-by: dbczumar <corey.zumar@databricks.com>

dbczumar committed:
fix

Signed-off-by: dbczumar <corey.zumar@databricks.com>

harupy reviewed (approved):
LGTM!

dbczumar committed:
schema fixes

Signed-off-by: dbczumar <corey.zumar@databricks.com>

dbczumar committed:
fix

Signed-off-by: dbczumar <corey.zumar@databricks.com>

dbczumar committed:
FIX

Signed-off-by: dbczumar <corey.zumar@databricks.com>

dbczumar committed:
fix

Signed-off-by: dbczumar <corey.zumar@databricks.com>

dbczumar committed:
fix

Signed-off-by: dbczumar <corey.zumar@databricks.com>
